### ISSUE TITLE
Report a bug about computing query norm in MIPS mode

### ISSUE THREAD

**Issue by HannanKan** [microsoft/DiskANN#16]
https://github.com/microsoft/DiskANN/blob/bc39cd06c0a1d668aaae9b97199d724438d617bf/src/pq_flash_index.cpp#L849
loop condition should be
`
 for (uint32_t i = 0; i < (this->data_dim-1); i++) { 
`
`this->data_dim` is assigned at
https://github.com/microsoft/DiskANN/blob/bc39cd06c0a1d668aaae9b97199d724438d617bf/src/pq_flash_index.cpp#L619

and `pq_file_dim` is assigned at 
https://github.com/microsoft/DiskANN/blob/bc39cd06c0a1d668aaae9b97199d724438d617bf/src/pq_flash_index.cpp#L608

`pq_file_dim` is equal to number of dimensions of vectors in index file. However, In MIPS, original vectors are transformed with an additional dimension, which is sqrt(1-||x||/||w||). So vectors in index file are 1 more dimension than original vectors, whose number of dimensions are equal to queries'.

**Comment by ShikharJ** [comment]
@HannanKan This is not a bug, when using DiskANN, both the input vectors and query vectors are expected to be of the same dimensionality. So the queries are expected to be padded with zeros for MIPS case.

**Comment by HannanKan** [comment]
I have a question. In building phase, each of input vectors has d dimensions. How many dimensions should a query have in search phase?

**Comment by ShikharJ** [comment]
Same as the input vector dimensions. In the regular case, it is just `d`. In MIPS, both become `d+1`.

**Comment by HannanKan** [comment]
Yeah. Both input of building phase and search phase should have same dimensions. But in building phases, input are changed to have an extra dimension. I debug with gdb and find that my query has actually `d` dimension but following codes read `d+1`.
https://github.com/microsoft/DiskANN/blob/bc39cd06c0a1d668aaae9b97199d724438d617bf/src/pq_flash_index.cpp#L849

**Comment by ShikharJ** [comment]
For MIPS, you'll need to use `tests/utils/vector_analysis` functionality to convert both input and query files, and then use them while building and searching.

**Comment by HannanKan** [comment]
In that case, the README may have to update.

**Comment by ShikharJ** [comment]
I agree, unfortunately, we're stretched thin at the moment. Thanks for bringing this to our notice, we appreciate it.

**Comment by HannanKan** [comment]
> For MIPS, you'll need to use `tests/utils/vector_analysis` functionality to convert both input and query files, and then use them while building and searching. Use `./vector_analysis float /path_to_data_file 4 /path_to_converted_data_file`.

I just read `tests/util/vector_analysis` codes. I think it may be used for small data because it load all data from disk into memory in one shot. Apart from that, if I choose mips when building index using `./test/build_disk_index`, input will be added an extra coordinate. An alternative way to use diskann without changing codes is to use `tests/util/vector_analysis` to convert query data.

**Comment by ShikharJ** [comment]
I see, let's leave this issue open for now.

**Comment by JingyuanHe1222** [comment]
Hi, as I looked at /src/pq_flash_index.cpp, the cached_beam_search() function seems to handle the extra dimension by 1. normalizing the queries; 2. make the last dimension 0 (the d+1 th dimension). In this case, do we still need to manually prepend an extra dimension and normalize the query vectors? 
Thanks in advance!
